src/gateways/GoPayRecurrent.php

```diff
@@ -66,42 +65,62 @@ public function notification(string $id, ?string $parentId = null): bool
             $payment->id
         );
 
-        if ($payment->status != PaymentsRepository::STATUS_PAID && in_array($data['state'], [self::STATE_PAID, self::STATE_AUTHORIZED])) {
-            if ((boolean) $payment->payment_gateway->is_recurrent) {
-                $recurrentPayment = $this->recurrentPaymentsRepository->findByPayment($payment);
-                if ($recurrentPayment) {
-                    $this->recurrentPaymentsProcessor->processChargedRecurrent(
-                        $recurrentPayment,
-                        PaymentsRepository::STATUS_PAID,
-                        $this->getResultCode(),
-                        $this->getResultMessage()
-                    );
-                } else {
-                    $payment = $this->paymentsRepository->updateStatus($payment, PaymentsRepository::STATUS_PAID, true);
-                    $this->recurrentPaymentsRepository->createFromPayment(
-                        $payment,
-                        $id
-                    );
-                }
-            } else {
-                $this->paymentsRepository->updateStatus($payment, PaymentsRepository::STATUS_PAID, true);
-            }
+        if ($payment->status !== PaymentsRepository::STATUS_FORM) {
+            return true;
+        }
+
+        if (empty($data['state']) && $this->getError() !== null) {
+            $this->handleCanceled($payment, PaymentsRepository::STATUS_FAIL);
+            return true;
+        }
+        $this->gopayPaymentsRepository->updatePayment($payment, $this->buildGopayPaymentValues($data));
+
+        if (in_array($data['state'], [self::STATE_PAID, self::STATE_AUTHORIZED])) {
+            $this->handleSuccess($payment);
         }
 
         if (in_array($data['state'], [self::STATE_CANCELED, self::STATE_TIMEOUTED])) {
-            if ((boolean) $payment->payment_gateway->is_recurrent) {
-                $recurrentPayment = $this->recurrentPaymentsRepository->findByPayment($payment);
-                if ($recurrentPayment) {
-                    $this->recurrentPaymentsProcessor->processFailedRecurrent($recurrentPayment, $this->getResultCode(), $this->getResultMessage());
-                } else {
-                    $this->paymentsRepository->updateStatus($payment, $data['state'] === self::STATE_TIMEOUTED ? PaymentsRepository::STATUS_TIMEOUT : PaymentsRepository::STATUS_FAIL, true);
-                }
+            $this->handleCanceled($payment, $data['state'] === self::STATE_TIMEOUTED ? PaymentsRepository::STATUS_TIMEOUT : PaymentsRepository::STATUS_FAIL);
+        }
+
+        return true;
```

Comment:
By explicitly checking `$data['state']` values above, there's a chance you might miss some unhandled status.

- At this point the handler doesn't do anything with `self::STATE_CREATED`. If it's correct behavior, I'd probably explicitly handle that state with its own `if` branch and call `return true`. As a reader, I'd have a clear idea that it's intentional.
- Basically each of those `if` branches checking `$data['state']` should probably end with `return` statement. There's no need to continue if the payment was already handled.
- If we get here, to the default `return true` at the end of method, it means that the notification contained a state that we didn't handle. I'd probably want to fail fast here and throw an exception with `unhandled state: $data['state]` message.